main.c

```diff
@@ -949,3 +959,126 @@ ErrorExit(int exit_code, const wchar_t *msg)
     }
 }
+
+/**
+ * Save a list of active connections in registry. If enable_auto_restart
+ * is false, any previously saved list is deleted and no new list is saved.
+ * Do not show any messages here as this may be called on receiving
+ * WM_ENDSESSION (user logout).
+ */
+static void
+SaveAutoRestartList()
+{
+    size_t len = 0;
+    int nactive = 0;
+    int max_active = o.num_configs - CountConnState(disconnected);
+
+    if (!o.enable_auto_restart || max_active <= 0)
+    {
+        /* delete the list -- on load this gets treated as an empty list */
+        RegDeleteKeyValueW(HKEY_CURRENT_USER, GUI_REGKEY_HKCU, L"auto_restart_list");
+        return;
+    }
+
+    int *active_conns = malloc((size_t) max_active*sizeof(int));
+    wchar_t *list = NULL;
+
+    if (!active_conns)
+    {
+        MsgToEventLog(EVENTLOG_ERROR_TYPE, L"Out of memory while persisting state in registry");
+        return;
+    }
+
+    for (int i = 0; i < o.num_configs; i++)
+    {
+        if (o.conn[i].state == disconnected
+            || o.conn[i].flags & FLAG_DAEMON_PERSISTENT)
+        {
+            continue;
+        }
+        /* accumulate space needed for list of active connections */
+        if (active_conns)
+        {
+            len += wcslen(o.conn[i].config_name) + 1;
+            active_conns[nactive++] = i;
+        }
+    }
+    len++; /* for double nul termination */
+
+    if (len == 1) len++; /* two nuls for empty string */
+
+    /* Make a double nul terminated list of active connections */
+    list = calloc(len, sizeof(wchar_t));
+    if (!list)
+    {
+        free(active_conns);
+        MsgToEventLog(EVENTLOG_ERROR_TYPE, L"Out of memory while persisting state in registry");
+        return;
+    }
+
+    wchar_t *p = list;
+    for (int i = 0; i < nactive; i++)
+    {
+        connection_t *c = &o.conn[active_conns[i]];
+        if (c->state == disconnected || c->flags & FLAG_DAEMON_PERSISTENT)
+        {
+            continue;
+        }
+        wcscpy(p, c->config_name); /* wcscpy is safe here */
+        p += wcslen(c->config_name) + 1;
+    }
+
+    /* Save the list in registry for auto-connect on restart */
+    LSTATUS status = RegSetKeyValueW(HKEY_CURRENT_USER, GUI_REGKEY_HKCU, L"auto_restart_list",
+                                   REG_MULTI_SZ, list, (DWORD) len*sizeof(wchar_t));
+    if (status != ERROR_SUCCESS)
+    {
+        MsgToEventLog(EVENTLOG_ERROR_TYPE, L"RegSetKeyValue returned error: status = %lu", status);
+    }
+
+    free(active_conns);
+    free(list);
+}
+
+/*
+ * Read list of active connections in last session and set them to auto-start
+ */
+static void
+LoadAutoRestartList()
+{
+    wchar_t *list;
+    DWORD len = 0;
+
+    LSTATUS status = RegGetValueW(HKEY_CURRENT_USER, GUI_REGKEY_HKCU, L"auto_restart_list",
+                                RRF_RT_REG_MULTI_SZ, NULL, NULL, &len);
+    if (status != ERROR_SUCCESS || len == 0)
+    {
+        return;
+    }
+
+    list = malloc(sizeof(len));
```

Comment:
Shouldn't it be `malloc(len)` ?

Reply:
That should be sizeof(wchar_t)*len
I did test this, couldn't have worked with such silly bugs. Let me check what I pushed.

Reply:
I managed to break it to heap corruption by putting long enough string into registry value. With short strings it just "works" - although I expected VS compiler to be smarter and detect those things.